src/Common/src/Interop/Unix/System.Native/Interop.readdir.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        internal enum NodeType : short
+        {
+            DT_UNKNOWN  =  0,
+            DT_FIFO     =  1,
+            DT_CHR      =  2,
+            DT_DIR      =  4,
+            DT_BLK      =  6,
+            DT_REG      =  8,
+            DT_LNK      = 10,
+            DT_SOCK     = 12,
+            DT_WHT      = 14
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct InternalDirectoryEntry
+        {
+            internal NodeType   InodeType;
+            internal IntPtr     NameStartAddress;
+            internal int        NameLength;
+        }
+
+        internal struct DirectoryEntry
+        {
+            internal NodeType   InodeType;
+            internal string     InodeName;
+        }
+
+        [DllImport(Libraries.SystemNative, SetLastError = true)]
+        internal static extern Microsoft.Win32.SafeHandles.SafeDirectoryHandle OpenDir(string path);
+
+        [DllImport(Libraries.SystemNative, SetLastError = false)]
+        internal static extern int GetDirentSize();
+
+        [DllImport(Libraries.SystemNative, EntryPoint = "ReadDir", SetLastError = true)]
+        private static extern int NativeReadDir(SafeDirectoryHandle dir, IntPtr buffer, out InternalDirectoryEntry outputEntry, out bool hasData);
+
+        [DllImport(Libraries.SystemNative, SetLastError = true)]
+        internal static extern int CloseDir(IntPtr directory);
+
+        // The calling pattern for ReadDir is described in src/Native/System.Native/pal_readdir.cpp
+        internal static bool ReadDir(Microsoft.Win32.SafeHandles.SafeDirectoryHandle dir, out DirectoryEntry outputEntry)
+        {
+            unsafe
+            {
+                // To reduce strcpys, alloc a buffer here and get the result from OS, then copy it over for the caller.
+                int bufferSize = GetDirentSize();
```

Comment:
Since this value never changes, it would be good to cache this. Otherwise, we have 2x P/Invoke cost for every ReadDir. I think it can be: `private static readonly int s_bufferSize = GetDirentSize();`